### ISSUE TITLE
There is no documentation available for Source Models, Backend Models, and Frontend Models

### ISSUE THREAD

**Issue by jonathanhodges** [magento/devdocs#365]
It would be great if there was documentation on what Source Models, Backend Models, and Frontend Models existed within default Magento.  I imagine this could be updated either manually or via a script by searching for classes that are not abstract and that implement the appropriate interface.  I expect that this would be the case for more than just these models including things like models available through the service layer.

**Comment by xcomSteveJohnson** [comment]
@tdobbs and @tanberry Please take a look.

**Comment by tanberry** [comment]
Thanks @jonathanhodges, great point! Little could you have known, but you have reawakened an old internal conversation... we very much want a script or tool that will generate such docs, but as with all things there is prioritization and workload balance to be determined.

Is there any chance that this is a project that you might be interested in helping out with?? We know that llamas are brilliant, and the Magento Devdocs team has an initiative to encourage and acknowledge contributions by our community members; everything from displaying the name of the contributor on the devdocs page, with link-back to the contributor's blog or company website, to a small travel budget to work on larger projects. Let us know!

**Comment by jonathanhodges** [comment]
So @tanberry, I was thinking about it this and I think that since you do not already have something like built, this can be done manually without too much difficulty.  The key is to have the appropriate interfaces documented.  Once that is done, anyone with an IDE can regenerate this list in a matter of minutes.  I think at least for the short term, the most effective way to solve this will be to document the Interfaces and periodically (every new version) go in and run the IDE command to find all classes which implement that interface and manually update that list.  This shouldn't take that much time and it will get the information in there much sooner.

In the long run, it may be helpful to reach out to a company like JetBrains who is already doing this sort of logic in their PHP product and see if they would be open to building something more robust that could be used for the generation of this information from a given codebase.

Regarding writing the documentation for this section.  I am open to it but I'm not sure when I'll be able to get to it.  I will check back next week and see if it still open and if I have time, I will get something for you.

**Comment by tanberry** [comment]
Thanks, @jonathanhodges, for the ideas and further explanation. I'll get with our internal team and see if we can get something done with the manual approach, (or long-term something more robust) but with 2.0 GA looming and everyone packed with work; it might not happen quickly. If it turns out you have any time to assist, that would be terrific, and we would make sure you got super-acknowledged!

**Comment by xcomSteveJohnson** [comment]
@jonathanhodges Can you elaborate on which interfaces should be documented? Thanks.

**Comment by jonathanhodges** [comment]
I think documentation on the interfaces that should be implemented to create Source Models, Backend Models, and Frontend Models as well as the exact format of the data that the methods should return and some full featured example implementations with walkthroughs would be ideal.  It would also be nice to have a list of available ones with a brief description of what that source model does and any custom how to use it information.  I think the last part is a nice to have but there have been times where I have implemented my own model only to find out later that core Magento already had one that I could have used.

Jonathan Hodges
Director of Development / Senior Engineer, Classy Llama
MAGENTO CERTIFIED DEVELOPER PLUS
o 417.866.8887 ext. 107 | m 518.893.1241 | jonathan@classyllama.com

**Comment by xcomSteveJohnson** [comment]
Created internal issue MAGETWO-48712

**Comment by jeff-matthews** [comment]
@dshevtsov, does the new "Module Reference Documentor" cover these areas (MAGETWO-38492)?

**Comment by jeff-matthews** [comment]
@dshevtsov, please reply.

**Comment by dshevtsov** [comment]
We plan to start using the Documentor to generate reference documentation for PHP code in `app/code/Magento` that is marked as`@api` code.

**Comment by dshevtsov** [comment]
There is a short use case of Models usage in the [How to Add a New Product Attribute](https://devdocs.magento.com/videos/fundamentals/add-new-product-attribute/) video

**Comment by shrielenee** [comment]
Hello! We are archiving this ticket. Our Magento Community and Core developers haven't had the bandwidth for an extended period of time to address this. We're archiving this in an effort to create a quality, community-driven backlog which will allow us to allocate the required attention more easily.

Please feel free to create a new issue if necessary. Thank you for continued collaboration and assistance!
